modules/codegen/src/smithy4s/codegen/internals/Renderer.scala

```diff
@@ -1644,60 +1644,76 @@ private[internals] class Renderer(compilationUnit: CompilationUnit) { self =>
-          def arrayNode(x: ArrayNode): Line = {
-            val innerValues = x.getElements().asScala.map(_.accept(this))
-            line"smithy4s.Document.array(${innerValues.toList.intercalate(Line.comma)})"
-          }
-          def booleanNode(x: BooleanNode): Line =
-            line"smithy4s.Document.fromBoolean(${x.getValue})"
-          def nullNode(x: NullNode): Line =
-            line"smithy4s.Document.nullDoc"
-          def numberNode(x: NumberNode): Line =
-            line"smithy4s.Document.fromDouble(${x.getValue.doubleValue()}d)"
-          def objectNode(x: ObjectNode): Line = {
-            val members = x.getMembers.asScala.map { member =>
-              val key = s""""${member._1.getValue()}""""
-              val value = member._2.accept(this)
-              line"$key -> $value"
+        ((ts: java.time.Instant) => line"${NameRef("smithy4s", "Timestamp")}(${ts.getEpochSecond()}L, ${ts.getNano()})")
+          .asInstanceOf[T => Line]
+      case Primitive.Document =>
+        { (node: Node) =>
+          node.accept(new NodeVisitor[Line] {
+            def arrayNode(x: ArrayNode): Line = {
+              val innerValues = x.getElements().asScala.map(_.accept(this))
+              line"smithy4s.Document.array(${innerValues.toList.intercalate(Line.comma)})"
             }
-            line"smithy4s.Document.obj(${members.toList.intercalate(Line.comma)})"
-          }
-          def stringNode(x: StringNode): Line =
-            line"""smithy4s.Document.fromString(${renderStringLiteral(
-              x.getValue
-            )})"""
-        })
-      }
-      case Primitive.Nothing => v => (v: Nothing) // this case can't happen
+            def booleanNode(x: BooleanNode): Line =
+              line"smithy4s.Document.fromBoolean(${x.getValue})"
+            def nullNode(x: NullNode): Line =
+              line"smithy4s.Document.nullDoc"
+            def numberNode(x: NumberNode): Line =
+              line"smithy4s.Document.fromDouble(${x.getValue.doubleValue()}d)"
+            def objectNode(x: ObjectNode): Line = {
+              val members = x.getMembers.asScala.map { member =>
+                val key = s""""${member._1.getValue()}""""
+                val value = member._2.accept(this)
+                line"$key -> $value"
+              }
+              line"smithy4s.Document.obj(${members.toList.intercalate(Line.comma)})"
+            }
+            def stringNode(x: StringNode): Line =
+              line"""smithy4s.Document.fromString(${renderStringLiteral(
+                x.getValue
+              )})"""
+          })
+        }.asInstanceOf[T => Line]
+      case Primitive.Nothing => _ => sys.error("cannot happen") // this case can't happen
     }
 
   private def renderStringLiteral(raw: String): Line = {
-    import scala.reflect.runtime.universe._
-    val str = Literal(Constant(raw))
-      .toString()
+    // Escape the string as a Scala literal (surround with quotes, escape special characters)
```

Comment:
homestly I think we should just use scalameta for small things like these (rendering constants)

Reply:
I tried it in the past and while having the snippets to compile at the same time you compile your code is mind-blowing there was a lot of hassle sometimes to get them right.

Reply:
yeah, the ancestor to smithy4s was actually written with scalameta. Lots of friction.